Manage snippet registry to write snippet in JSON

angelozerr opened the pull request:
Fixes #640

Signed-off-by: azerr <azerr@redhat.com>

angelozerr commented:
This PR manages snippets with vscode JSON format. In other words

 * existing snippets:

  * CDATA snippets
  * Comments snippets
  * DTD ELEMENT, ENTITY, ATTLIST snippets (inside DTD or DOCTYPE)

 * new snippets for doctype.

Here some advanced snippet support:

 * snippet are displayed according some condition (ex : DOCTYPE snippet is working only if file is empty or if completion is trigger before the root element which doesn't declare a DOCTPE.
 * when DOCTYPE snippet is applied, it uses teh root element name to generate the proper DOCTYPE.

angelozerr commented:
Here a demo:

![DocTypeSnippetDemo](https://user-images.githubusercontent.com/1932211/81189037-5f112380-8fb6-11ea-8e94-349213c3a572.gif)

angelozerr commented:
@fbricon @dakwon you can play now with my PR, I need to clean my code and add some other tests.

fbricon commented:
xml prolog snippet should be the 1st choice in an empty xml document. It's totally missing

fbricon commented:
xml prolog should also probably be added to all snippets

fbricon commented:
If there's an xml prolog already, the doctype snippets are not even proposed on the next line

xorye commented:
I'm very excited for this PR.

For the prolog snippet named "Insert xml Processing Instruction with standalone", one of the placeholders are appearing:
![image](https://user-images.githubusercontent.com/20326645/81226797-684fbf80-8fb9-11ea-9637-f845599ad1f3.png)

Also, I think the space before the `?>` at the end should be gone?

angelozerr commented:
> xml prolog should also probably be added to all snippets

done

> If there's an xml prolog already, the doctype snippets are not even proposed on the next line

fixed

>For the prolog snippet named "Insert xml Processing Instruction with standalone", one of the placeholders are appearing:

fixed

> Also, I think the space before the ?> at the end should be gone?

fixed

@xorye could you play again with this PR and give me feedback, thanks!

angelozerr commented:
> xml prolog snippet should be the 1st choice in an empty xml document. It's totally missing

I have not managed sort for te moment, could we see that in an another PR please.

angelozerr commented:
Code should be cleaned and test are writen.

angelozerr commented:
Here a new demo:

![SnippetDemo](https://user-images.githubusercontent.com/1932211/81324222-ef269a00-9096-11ea-8665-1394f0e085e2.gif)

xorye commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/snippets/SnippetRegistry.java:
```diff
@@ -0,0 +1,380 @@
+	 * @param contextFilter      the context filter.
+	 * @return the snippet completion items according to the context filter.
+	 */
+	public List<CompletionItem> getCompletionItems(Range replaceRange, String lineDelimiter, boolean canSupportMarkdown,
```
I think after #680 and #669 are both merged, we should use `xml.preferences.quoteStyle` to determine whether or not to use `"` or `'` in the insert snippet text. I can create an issue for that.

angelozerr replied:
Yes please create an issue.

xorye commented:
The features in the demo gif work for me.

I have some feedback:

When I get completion after typing `<?`, do we still need the `<?xml ... ?>` option?
![image](https://user-images.githubusercontent.com/20326645/81342755-c1832600-9081-11ea-93a5-2309a4b6056c.png)

Also, when I select `Insert xml Processing Instruction` or `Insert xml Processing Instruction with standalone`, I think there is a problem with the replace range:
![image](https://user-images.githubusercontent.com/20326645/81342951-19ba2800-9082-11ea-91c8-0a1b16c4f76f.png)
There is an extra `<?`. I only get this problem with the prolog snippets.

Also if completion is invoked inside a comment like this:
![image](https://user-images.githubusercontent.com/20326645/81344048-0f992900-9084-11ea-92fd-14ec4244e645.png)

I don't think we should be getting any snippets in this case

xorye commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/snippets/PrologSnippetContext.java:
```diff
@@ -0,0 +1,60 @@
+		}
+		// No xml processing instruction, check if completion was triggered before the
+		// document element
+		DOMElement documentElement = document.getDocumentElement();
```
Can we also check for DocType? If I invoke completion at `<!DOCTYPE |root-element [`, I don't think I should get the prolog snippets.

![image](https://user-images.githubusercontent.com/20326645/81344693-458add00-9085-11ea-96b4-04b1ad052496.png)

fbricon commented:
Please fix the merge conflict

angelozerr commented on org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json:
```diff
@@ -0,0 +1,53 @@
+      "\t<!ELEMENT ${1:root-element} (${2:#PCDATA}) >",
+      "]>"
+    ],
+    "description": "Insert  DOCTYPE with subset"
```
@xorye fixed

angelozerr commented:
> xml prolog should also probably be added to all snippets

It can be complicated to do that because we need to manage case of existing or not of xml processing instruction to avoid generating twice.

I had the problem with DOCTYPE (which generates a root-element if doens't exist and don't generate it otherwise). To do that I have duplicated snippets (see newfile-snippets.json and doctype-snippets.json).

To manage properly this usecase we should manage aggregation of several snippets, but perhaps we could do that in the future? 

> Also, when I select Insert xml Processing Instruction or Insert xml Processing Instruction with standalone, I think there is a problem with the replace range:
There is an extra <?. I only get this problem with the prolog snippets.

@xorye fixed

> Also if completion is invoked inside a comment like this:
I don't think we should be getting any snippets in this case

@xorye fixed

> Can we also check for DocType? If I invoke completion at <!DOCTYPE |root-element [, I don't think I should get the prolog snippets.

@xorye fixed

> Please fix the merge conflict

@fbricon  fixed

The JSON snippet needs to be improved again, but if you are OK could we done that in separate issues:

 * gives the capability to sort the snippets (xml prolog snippet should be the 1st choice in an empty xml document. )
 * eat the '>' when it exists. I mean when we have `<? | >` and xml processing instruction is applied, it doesn't eat the `>` and it generates

```xml
<?xml version="1.0" encoding="UTF-8"?>>
```
 * snippet should take care of quote preferences.

xorye commented:
Great, the fixes work on my machine.

I found a small inconsistency, XML should either always be upper case or always be lower case. 
![image](https://user-images.githubusercontent.com/20326645/81600888-55adff80-9398-11ea-95db-396a32b05100.png)
![image](https://user-images.githubusercontent.com/20326645/81600922-6199c180-9398-11ea-8b0b-f35719640068.png)

xorye commented on org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json:
```diff
@@ -0,0 +1,58 @@
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} [",
+      "\t<!ELEMENT ${1:root-element} (${2:#PCDATA}) >",
```
Remove extra space before the closing `>`

angelozerr replied:
fixed

xorye commented on org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json:
```diff
@@ -0,0 +1,58 @@
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} PUBLIC \"${2:public-id}\" \"${3:file}.dtd\" >"
```
Remove extra space before the closing `>`

angelozerr replied:
fixed

xorye commented on org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json:
```diff
@@ -0,0 +1,97 @@
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} SYSTEM \"${2:file}.dtd\" >",
```
Remove extra space before the closing `>`

angelozerr replied:
fixed

xorye commented on org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json:
```diff
@@ -0,0 +1,97 @@
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} PUBLIC \"${2:public-id}\" \"${3:file}.dtd\" >",
```
Remove extra space before the closing `>`

angelozerr replied:
fixed

xorye commented on org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json:
```diff
@@ -0,0 +1,97 @@
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} [",
+      "\t<!ELEMENT ${1:root-element} (${2:#PCDATA}) >",
```
Remove extra space before the closing `>`

angelozerr replied:
fixed

xorye commented:
Also, the completion options are a bit hard to identify IMO:
![image](https://user-images.githubusercontent.com/20326645/81601640-35327500-9399-11ea-806e-d4b73fed2b41.png)

I think the old descriptions were nicer:
![image](https://user-images.githubusercontent.com/20326645/81601672-44b1be00-9399-11ea-96c2-ec275654acbd.png)

xorye commented:
For the noNamespaceSchemaLocation snippet, the attributes are on a separate line:
```
<root-element
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:noNamespaceSchemaLocation="file.xsd">
    
</root-element>
```

A cool enhancement (maybe this can be done on a different issue) would be to have the attributes on a separate line if `xml.format.splitAttributes` is true. 

Taking it one step further, it might be cool to have snippet text be preformatted against the formatter. 
Because if the user has format on save enabled, and the user saves the document after inserting a snippet, I think it would be weird if the snippet content's formatting changes dramatically (like in the case for split attributes)

angelozerr commented:
> I found a small inconsistency, XML should either always be upper case or always be lower case.

fixed

angelozerr commented:
> But things like this could be discussed in another issue.

I agree with you we should pay attention with format settings in otherwise snippet should be formatted (each time or better one time and cache the formatted snippet).

Please create an issue for that. Thanks

fbricon commented:
Jenkins fails https://ci.eclipse.org/lemminx/blue/organizations/jenkins/lemminx/detail/PR-680/34/pipeline:
```
[ERROR] Failures: 

[ERROR]   XMLCompletionSnippetsTest.afterComment:118 expected: <<!DOCTYPE root-element SYSTEM "file.dtd" >

<root-element>

</root-element>> but was: <<!DOCTYPE root-element SYSTEM "file.dtd">

<root-element>

</root-element>>

[ERROR]   XMLCompletionSnippetsTest.afterProlog:133 expected: <<!DOCTYPE root-element SYSTEM "file.dtd" >

<root-element>

</root-element>> but was: <<!DOCTYPE root-element SYSTEM "file.dtd">

<root-element>

</root-element>>

[ERROR]   XMLCompletionSnippetsTest.emptyXMLContent:38 expected: <<!DOCTYPE root-element SYSTEM "file.dtd" >

<root-element>

</root-element>> but was: <<!DOCTYPE root-element SYSTEM "file.dtd">

<root-element>

</root-element>>
```

angelozerr commented:
@fbricon fixed

xorye commented:
The snippet is eating more than the `>`:

![gif](https://raw.githubusercontent.com/xorye/gifs/master/new_issues/comment.gif?token=AE3CR5KZPTE3BJURUK6ZOAS6YPIMA)

If this is not a big problem, I'm good with this PR being merged

angelozerr commented:
> If this is not a big problem, I'm good with this PR being merged

No it's very bad -( Let me try to fix it.
